draft-ietf-jsonpath-base.md

```diff
@@ -1402,91 +1454,125 @@ According to {{filter-selector}}, a `function-expr` is valid as a `filter-path`
   in a test expression or as a function argument).
+  Members of `NodesType` have no syntactical representation in JSON Path.
 
 The abstract instances above can be obtained from the concrete representations in {{tbl-typerep}}.
 
 | Abstract Instance | Concrete Representations                                           |
 | :---------------: | :----------------------:                                           |
-| `Node(n)`         | Singular Path resulting in a nodelist containing just the node `n` |
 | `Value(v)`        | JSON value `v`                                                     |
-| `Nothing`         | Singular Path or `filter-path` resulting in an empty nodelist      |
-| `Nodes(nl)`       | `filter-path` resulting in the non-empty nodelist `nl`             |
+| `Nothing`         | A representation of the absence of a JSON value, distinct from the JSON literal `null`, e.g., from a Singular Path or `filter-path` resulting in an empty nodelist   |
+| `Nodes(nl)`       | A list of zero or more nodes, e.g., from a `filter-path` resulting in the nodelist `nl`, which may or may not be empty  |
 {: #tbl-typerep title="Concrete representations of abstract instances"}
 
-The following subtype relationships depend on coercion:
-
-* `OptionalNodeType` is a subtype of `OptionalValueType` via coercion since the `OptionalNodeType` instance `Node(n)` can be coerced to
-the `OptionalValueType` instance `Value(v)`, where `v` is the value of the node `n`.
-* `OptionalNodeType` is a subtype of `OptionalNodesType` via coercion since the `OptionalNodeType` instance `Node(n)` can be coerced to
-the `OptionalNodesType` instance `Nodes(l)`, where `l` is a nodelist consisting of just the node `n`.
+### Type Conversions {#type-conv}
+
+The following type conversions may occur:
+
+* Where an expression with a declared type of `NodesType` needs to be
+  converted to a `ValueType`, the conversion proceeds as follows:
+  * If the nodelist contains a single node, the conversion result is
+    the value of the node.
+  * If the nodelist is empty or contains multiple nodes, the
+    conversion result is `Nothing`.
+* Where a member of `NodesType` needs to be converted to a
+  `LogicalType`, the conversion proceeds as follows:
+  * If the nodelist contains one or more nodes, the conversion result
+    is `LogicalTrue`.
+  * If the nodelist is empty, the conversion result is `LogicalFalse`.
+
+The well-typedness of function expressions can now be defined in terms of this type system.
+
+### Well-Typedness of Function Expressions
+
+A function expression is well-typed if all of the following are true:
+
+* If it occurs directly in a test expression, the function is declared
+  to have a result type of `LogicalType`, or (conversion applies)
+  `NodesType`.
+* If it occurs directly as a `comparable` in a comparison, the
+  function is declared to have a result type of `ValueType`, or
+  (conversion applies) `NodesType`.
```

Comment:
If you want to do this, you'd have to define a new type for single-nodelists (separate from `NodesType`) so that functions can specify that this is what they return and can be used in comparisons.  While you could then restrict `NodesType` functions to only appearing in test expressions, I'm not convinced this new type is any more useful than the function just returning the value that the node contains (`ValueType`).